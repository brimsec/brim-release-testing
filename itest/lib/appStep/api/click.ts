import {Application} from "spectron"

import {retryUntil} from "../../control"
import {LOG} from "../../log"

import logStep from "../util/logStep"

const waitForClickable = async (app: Application, selector: string) => {
  // In testing, it's been shown than there is no need to scroll to
  // elements to make them be visible, as long as the element's
  // container allows scrolling. However, some Internet searches suggest
  // scrolling to the element before trying to click in order to avoid
  // problems like those described in
  // https://github.com/brimsec/brim/issues/668
  await logStep(`wait for element to exist: "${selector}"`, () =>
    app.client.waitForExist(selector)
  )
  await logStep(`wait for element to be visible: "${selector}"`, () =>
    app.client.waitForVisible(selector)
  )
  return logStep(`scroll to: "${selector}"`, () => app.client.scroll(selector))
}

export const click = (app: Application, selector: string) =>
  logStep(`click on selector "${selector}"`, async () => {
    await waitForClickable(app, selector)
    try {
      return retryUntil(
        () => app.client.click(selector),
        (success) => success
      )
    } catch (e) {
      LOG.debug("trying to execute script for click: " + e)
      return app.client.selectorExecute(selector, (elem) => {
        elem.click()
      })
    }
  })

export const rightClick = (app: Application, selector: string) =>
  logStep(`right-click on selector "${selector}"`, async () => {
    await waitForClickable(app, selector)
    try {
      return retryUntil(
        () => app.client.rightClick(selector),
        (success) => success
      )
    } catch (e) {
      LOG.debug("trying to execute script for rightClick: " + e)
      return app.client.selectorExecute(selector, (elem) => {
        elem.rightClick()
      })
    }
  })

export const waitForClickableButtonAndClick = async (
  app: Application,
  selector: string
) => {
  await waitForClickable(app, selector)
  // In addition to waitForClickable above, buttons must also be enabled.
  await logStep(`wait for button ${selector} to be enabled`, () =>
    retryUntil(
      () => app.client.getAttribute(selector, "disabled"),
      (isDisabled) => !isDisabled
    )
  )

  // We can use app.client.click() here because we've done the necessary
  // waiting.
  try {
    return retryUntil(
      () => app.client.click(selector),
      (success) => success
    )
  } catch (e) {
    LOG.debug("trying to execute script for click: " + e)
    return app.client.selectorExecute(selector, (elem) => {
      elem.click()
    })
  }
}
