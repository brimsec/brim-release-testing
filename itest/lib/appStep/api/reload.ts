import {Application} from "spectron"

import logStep from "../util/logStep"
import {retry} from "../../control"

export default (app: Application) =>
  // Issue a browser window reload(). This is flaky in the Spectron,
  // Chromedriver stack, but appears be contained within the method.
  // It seems safe to retry.
  // https://github.com/brimsec/brim/issues/878
  // https://github.com/electron-userland/spectron/issues/493
  logStep("app reload", () => retry(() => app.browserWindow.reload(), 60, 1000))
