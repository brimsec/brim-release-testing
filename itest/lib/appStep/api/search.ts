import {Application} from "spectron"

import logStep from "../util/logStep"
import waitForResults from "../util//waitForResults"
import {selectors} from "../../../../src/js/test/integration"

export const runSearch = async (app: Application, searchText: string) => {
  // Run a search and wait for results to appear. Do not return results.
  // This is suitable for setting up the app to do something else
  // after receiving results, like a context menu test.
  await logStep("write to main search", () =>
    app.client
      .waitForVisible(selectors.search.input)
      .then(() => app.client.setValue(selectors.search.input, searchText))
  )
  await logStep("click the search button", () => app.client.keys("Enter"))
  return waitForResults(app)
}

export const getResults = async (
  app: Application,
  includeHeaders = true
): Promise<string[]> => {
  // Return search current search results. This is useful for getting
  // the current results by doing a search through other means, like a
  // context menu-built search.

  const searchDisplayHeaders = async (app: Application): Promise<string[]> => {
    // This stinks. We have to use getHTML because headers that are off the
    // screen return as empty strings if you use getText. This isn't required of
    // actual results.
    // See http://v4.webdriver.io/api/property/getText.html
    // app.browserWindow.maximize() fixes the problem on my laptop but not CI.
    // But what we get back includes the width which can be non-deterministic:
    // <div class="header-cell" style="width: 192px;">ts<div class="col-resizer"></div></div>
    // That style width will vary on my laptop vs. CI.
    // The hack is to split this and extract just the text.
    const _trim = (s: string) => s.split(">")[1].split("<")[0]

    let headers = await logStep("get search headers", () =>
      app.client
        .waitForVisible(selectors.viewer.headers)
        .then(() => app.client.getHTML(selectors.viewer.headers))
    )
    if (typeof headers === "string") {
      headers = [headers]
    }
    return headers.map((h) => _trim(h))
  }

  const searchResults = await logStep("get search records", async () => {
    await app.client.waitForVisible(selectors.viewer.results)
    return app.client.getText(selectors.viewer.results)
  })

  let headers
  if (
    includeHeaders &&
    (await app.client.isVisible(selectors.viewer.headers))
  ) {
    headers = await searchDisplayHeaders(app)
  } else {
    headers = []
  }
  return headers.concat(searchResults)
}

export default async (app: Application, searchText: string) => {
  // Run a search and return results.
  await runSearch(app, searchText)
  return getResults(app)
}
