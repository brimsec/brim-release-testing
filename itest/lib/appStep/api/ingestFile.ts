import {Application} from "spectron"
import path from "path"

import {testDataDir} from "../../env"
import {retryUntil} from "../../control"
import {click} from "./click"
import logStep from "../util/logStep"
import {LOG} from "../../log"
import {selectors} from "../../../../src/js/test/integration"

export default async (app: Application, file: string) => {
  // Ingest a file inside the itest/testdata directory
  await click(app, ".add-tab")

  await logStep("wait for pcap file input", () =>
    app.client.waitForVisible(selectors.ingest.filesButton)
  )
  await logStep("choose file", () =>
    app.client.chooseFile(
      selectors.ingest.filesInput,
      path.normalize(path.join(testDataDir(), file))
    )
  )

  try {
    await logStep("wait for ingest to start", () =>
      retryUntil(
        () => app.client.isExisting(selectors.status.ingestProgress),
        (ingesting) => ingesting === true,
        100,
        100
      )
    )
  } catch {
    LOG.debug(
      "ingest never appeared; let's hope it finished and ended before we could observe it"
    )
  }

  return logStep("wait for ingest to finish", () =>
    retryUntil(
      () => app.client.isExisting(selectors.status.ingestProgress),
      (ingesting) => ingesting === false
    )
  )
}
