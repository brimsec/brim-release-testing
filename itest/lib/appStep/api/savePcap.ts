import {Application} from "spectron"

import {retryUntil} from "../../control"
import {waitForClickableButtonAndClick} from "./click"
import logStep from "../util/logStep"
import {selectors} from "../../../../src/js/test/integration"

export default async (app: Application) => {
  await waitForClickableButtonAndClick(app, selectors.pcaps.button)
  return logStep("wait for a download to finish", async () => {
    await app.client.waitForVisible(selectors.downloadMessage)
    return retryUntil(
      () => app.client.getText(selectors.downloadMessage),
      (text) => text == "Download Complete" || text.includes("Download error")
    )
  })
}

export const pcapsDir = (app: Application) =>
  app.electron.remote.app.getPath("temp")
