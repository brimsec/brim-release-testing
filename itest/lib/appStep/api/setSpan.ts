import {Application} from "spectron"

import {click} from "./click"
import logStep from "../util/logStep"
import waitForResults from "../util/waitForResults"
import {selectors} from "../../../../src/js/test/integration"

export default async (app: Application, span: string) => {
  // Set the span given by "string", like "Whole Space". This waits for
  // results to be redrawn before returning.
  await logStep("click span selector", () => click(app, selectors.span.button))
  await logStep(`select span ${span}`, () =>
    click(app, selectors.span.menuItem(span))
  )
  return waitForResults(app)
}
