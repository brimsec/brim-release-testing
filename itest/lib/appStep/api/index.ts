import {click, rightClick} from "./click"
import ingestFile from "./ingestFile"
import reload from "./reload"
import savePcap from "./savePcap"
import search from "./search"
import setSpan from "./setSpan"
import startApp from "./startApp"
import takeScreenshot from "./takeScreenshot"

export default {
  click,
  rightClick,
  ingestFile,
  reload,
  savePcap,
  search,
  setSpan,
  startApp,
  takeScreenshot
}
