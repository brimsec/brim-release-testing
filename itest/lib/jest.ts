import {Application} from "spectron"

import appStep from "./appStep/api"
import {selectors} from "../../src/js/test/integration"
import {LOG} from "./log"

export const TestTimeout = 300000
// https://jestjs.io/docs/en/troubleshooting#unresolved-promises
// https://jestjs.io/docs/en/jest-object#jestsettimeouttimeout
jest.setTimeout(TestTimeout)

export const handleError = async (
  app: Application,
  initialError: Error,
  done: any
) => {
  appStep.takeScreenshot(app)
  let realError = undefined
  let notificationError = undefined
  LOG.error(`handleError: Test hit exception: ${initialError.message}`)
  LOG.info("handleError: Looking for any desktop app notifications")
  try {
    notificationError = await app.client.getHTML(selectors.notification, false)
  } catch (e) {
    notificationError = undefined
  }
  if (notificationError) {
    realError = new Error(
      "App notification '" +
        notificationError +
        "' (initial error: '" +
        initialError +
        "'"
    )
  } else {
    LOG.info("handleError: desktop app notification not found")
    realError = initialError
  }
  done.fail && done.fail(realError)
}

// Put tests in a wrapper that prints what test is being run. This is
// to eagerly print to the console where the run is so that hangs and other
// failures are easier to diagnose. Without this, Jest is pretty silent and
// it's hard to know what went wrong in a CI run. This MUST be paired with jest
// --verbose to work properly, otherwise LOG.info() calls seem to get
// buffered and printed later.
export const stdTest = (
  descr: string,
  f: (done: any) => void | Promise<void>,
  timeout: number = TestTimeout
) => {
  test(
    descr,
    (done) => {
      LOG.info(`Running test: ${descr}`)
      f(done)
    },
    timeout
  )
}
