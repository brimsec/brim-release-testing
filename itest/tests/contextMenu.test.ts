import {basename} from "path"
import {sprintf} from "sprintf-js"

import {getResults, runSearch} from "../lib/appStep/api/search"
import appStep from "../lib/appStep/api"
import newAppInstance from "../lib/newAppInstance"
import {handleError, stdTest} from "../lib/jest"
import {selectors} from "../../src/js/test/integration"

// Fields we can test in types.tsv
const FIELDS = ["scalar", "record.scalar"]
// Brim character that represents an unset zng value. It must be
// handled specially as it can be a string value or the unset
// representation. That representation is only used to find a cell, or
// to exclude the actual character from search results.
const UNSET = "⦻"
const STRINGS = [
  "mystr",
  "\\x2d",
  "\\x09",
  '"',
  "'",
  ",",
  ";",
  "∫£œßü™",
  '"mystr"',
  "'mystr'",
  "*",
  '"*"',
  "1.1.1.1",
  " ",
  "null",
  "⦻" // The actual character, not unset.
]
const ADDRS = ["1.1.1.1", "fe80::58d2:2d09:e8cb:a8ad", "::"]

describe("type-wise Filter = value searches", () => {
  let app

  beforeAll(async (done) => {
    app = newAppInstance(basename(__filename) + "-types", 0)
    await appStep.startApp(app)
    await appStep.ingestFile(app, "types.tsv")
    await appStep.setSpan(app, "Whole Space")
    done()
  })

  afterAll(() => {
    if (app && app.isRunning()) {
      return app.stop()
    }
  })

  stdTest("FilterEq ensure ingest", (done) => {
    appStep
      .search(app, `* | count() by _path | sort _path`)
      .then((results) => {
        expect(results).toMatchSnapshot()
        done()
      })
      .catch((err) => {
        handleError(app, err, done)
      })
  })

  const run = (path: string, values: string[]) => {
    let testIdx = 0
    FIELDS.forEach((fieldName) => {
      values.forEach((s) => {
        const testId = sprintf("%04d", testIdx++)
        stdTest(`FilterEq${testId}: ${path} ${fieldName}="${s}"`, (done) => {
          runSearch(
            app,
            `_path=${path} ${fieldName}!=null | cut id, ${fieldName}`
          )
            .then(async () => {
              await appStep.rightClick(
                app,
                selectors.viewer.resultCellContaining(s)
              )
              await appStep.click(
                app,
                selectors.viewer.contextMenuItem("Filter = value")
              )
              expect(await getResults(app)).toMatchSnapshot()
              done()
            })
            .catch((err) => {
              handleError(app, err, done)
            })
        })
      })
    })
  }
  run("string", STRINGS)

  stdTest(`FilterEq unset/${UNSET} string`, (done) => {
    runSearch(app, `_path=string | cut id, scalar | sort -r id | head 10`)
      .then(async () => {
        await appStep.rightClick(
          app,
          selectors.viewer.resultCellContaining(UNSET)
        )
        await appStep.click(
          app,
          selectors.viewer.contextMenuItem("Filter = value")
        )
        expect(await getResults(app)).toMatchSnapshot()
        done()
      })
      .catch((err) => {
        handleError(app, err, done)
      })
  })

  run("addr", ADDRS)
})
