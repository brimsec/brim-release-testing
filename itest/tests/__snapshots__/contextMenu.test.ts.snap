// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`type-wise Filter = value searches FilterEq ensure ingest 1`] = `
Array [
  "_path",
  "count",
  "addr",
  "11",
  "string",
  "48",
]
`;

exports[`type-wise Filter = value searches FilterEq unset/⦻ string 1`] = `
Array [
  "id",
  "scalar",
  "61",
  "⦻",
  "60",
  "⦻",
  "59",
  "⦻",
  "58",
  "⦻",
  "53",
  "⦻",
  "52",
  "⦻",
  "51",
  "⦻",
  "50",
  "⦻",
  "49",
  "⦻",
  "48",
  "⦻",
]
`;

exports[`type-wise Filter = value searches FilterEq0000: addr scalar="1.1.1.1" 1`] = `
Array [
  "id",
  "scalar",
  "1",
  "1.1.1.1",
]
`;

exports[`type-wise Filter = value searches FilterEq0000: string scalar="mystr" 1`] = `
Array [
  "id",
  "scalar",
  "14",
  "mystr",
]
`;

exports[`type-wise Filter = value searches FilterEq0001: addr scalar="fe80::58d2:2d09:e8cb:a8ad" 1`] = `
Array [
  "id",
  "scalar",
  "3",
  "fe80::58d2:2d09:e8cb:a8ad",
]
`;

exports[`type-wise Filter = value searches FilterEq0001: string scalar="\\x2d" 1`] = `
Array [
  "id",
  "scalar",
  "16",
  "\\\\x2d",
]
`;

exports[`type-wise Filter = value searches FilterEq0002: addr scalar="::" 1`] = `
Array [
  "id",
  "scalar",
  "2",
  "::",
]
`;

exports[`type-wise Filter = value searches FilterEq0002: string scalar="\\x09" 1`] = `
Array [
  "id",
  "scalar",
  "17",
  "\\\\x09",
]
`;

exports[`type-wise Filter = value searches FilterEq0003: addr record.scalar="1.1.1.1" 1`] = `
Array [
  "id",
  "record.scalar",
  "5",
  "1.1.1.1",
]
`;

exports[`type-wise Filter = value searches FilterEq0003: string scalar=""" 1`] = `
Array [
  "id",
  "scalar",
  "18",
  "\\"",
]
`;

exports[`type-wise Filter = value searches FilterEq0004: addr record.scalar="fe80::58d2:2d09:e8cb:a8ad" 1`] = `
Array [
  "id",
  "record.scalar",
  "7",
  "fe80::58d2:2d09:e8cb:a8ad",
]
`;

exports[`type-wise Filter = value searches FilterEq0004: string scalar="'" 1`] = `
Array [
  "id",
  "scalar",
  "19",
  "'",
]
`;

exports[`type-wise Filter = value searches FilterEq0005: addr record.scalar="::" 1`] = `
Array [
  "id",
  "record.scalar",
  "6",
  "::",
]
`;

exports[`type-wise Filter = value searches FilterEq0005: string scalar="," 1`] = `
Array [
  "id",
  "scalar",
  "20",
  ",",
  "21",
  ",",
]
`;

exports[`type-wise Filter = value searches FilterEq0006: string scalar=";" 1`] = `
Array [
  "id",
  "scalar",
  "22",
  ";",
]
`;

exports[`type-wise Filter = value searches FilterEq0007: string scalar="∫£œßü™" 1`] = `
Array [
  "id",
  "scalar",
  "23",
  "∫£œßü™",
]
`;

exports[`type-wise Filter = value searches FilterEq0008: string scalar=""mystr"" 1`] = `
Array [
  "id",
  "scalar",
  "24",
  "\\"mystr\\"",
]
`;

exports[`type-wise Filter = value searches FilterEq0009: string scalar="'mystr'" 1`] = `
Array [
  "id",
  "scalar",
  "25",
  "'mystr'",
]
`;

exports[`type-wise Filter = value searches FilterEq0010: string scalar="*" 1`] = `
Array [
  "id",
  "scalar",
  "26",
  "*",
]
`;

exports[`type-wise Filter = value searches FilterEq0011: string scalar=""*"" 1`] = `
Array [
  "id",
  "scalar",
  "27",
  "\\"*\\"",
]
`;

exports[`type-wise Filter = value searches FilterEq0012: string scalar="1.1.1.1" 1`] = `
Array [
  "id",
  "scalar",
  "28",
  "1.1.1.1",
]
`;

exports[`type-wise Filter = value searches FilterEq0013: string scalar=" " 1`] = `
Array [
  "id",
  "scalar",
  "54",
  "",
  "55",
  "",
]
`;

exports[`type-wise Filter = value searches FilterEq0014: string scalar="null" 1`] = `
Array [
  "id",
  "scalar",
  "57",
  "null",
]
`;

exports[`type-wise Filter = value searches FilterEq0015: string scalar="⦻" 1`] = `
Array [
  "id",
  "scalar",
  "56",
  "⦻",
]
`;

exports[`type-wise Filter = value searches FilterEq0016: string record.scalar="mystr" 1`] = `
Array [
  "id",
  "record.scalar",
  "29",
  "mystr",
]
`;

exports[`type-wise Filter = value searches FilterEq0017: string record.scalar="\\x2d" 1`] = `
Array [
  "id",
  "record.scalar",
  "31",
  "\\\\x2d",
]
`;

exports[`type-wise Filter = value searches FilterEq0018: string record.scalar="\\x09" 1`] = `
Array [
  "id",
  "record.scalar",
  "32",
  "\\\\x09",
]
`;

exports[`type-wise Filter = value searches FilterEq0019: string record.scalar=""" 1`] = `
Array [
  "id",
  "record.scalar",
  "33",
  "\\"",
]
`;

exports[`type-wise Filter = value searches FilterEq0020: string record.scalar="'" 1`] = `
Array [
  "id",
  "record.scalar",
  "34",
  "'",
]
`;

exports[`type-wise Filter = value searches FilterEq0021: string record.scalar="," 1`] = `
Array [
  "id",
  "record.scalar",
  "35",
  ",",
  "36",
  ",",
]
`;

exports[`type-wise Filter = value searches FilterEq0022: string record.scalar=";" 1`] = `
Array [
  "id",
  "record.scalar",
  "37",
  ";",
]
`;

exports[`type-wise Filter = value searches FilterEq0023: string record.scalar="∫£œßü™" 1`] = `
Array [
  "id",
  "record.scalar",
  "38",
  "∫£œßü™",
]
`;

exports[`type-wise Filter = value searches FilterEq0024: string record.scalar=""mystr"" 1`] = `
Array [
  "id",
  "record.scalar",
  "39",
  "\\"mystr\\"",
]
`;

exports[`type-wise Filter = value searches FilterEq0025: string record.scalar="'mystr'" 1`] = `
Array [
  "id",
  "record.scalar",
  "40",
  "'mystr'",
]
`;

exports[`type-wise Filter = value searches FilterEq0026: string record.scalar="*" 1`] = `
Array [
  "id",
  "record.scalar",
  "41",
  "*",
]
`;

exports[`type-wise Filter = value searches FilterEq0027: string record.scalar=""*"" 1`] = `
Array [
  "id",
  "record.scalar",
  "42",
  "\\"*\\"",
]
`;

exports[`type-wise Filter = value searches FilterEq0028: string record.scalar="1.1.1.1" 1`] = `
Array [
  "id",
  "record.scalar",
  "43",
  "1.1.1.1",
]
`;

exports[`type-wise Filter = value searches FilterEq0029: string record.scalar=" " 1`] = `
Array [
  "id",
  "record.scalar",
  "59",
  "",
]
`;

exports[`type-wise Filter = value searches FilterEq0030: string record.scalar="null" 1`] = `
Array [
  "id",
  "record.scalar",
  "61",
  "null",
]
`;

exports[`type-wise Filter = value searches FilterEq0031: string record.scalar="⦻" 1`] = `
Array [
  "id",
  "record.scalar",
  "60",
  "⦻",
]
`;
